discovery/ecs/client/aws.go

```diff
@@ -0,0 +1,541 @@
+			cInstances, err := c.getContainerInstances(&cluster)
+			if err != nil {
+				globErr = err
+				return
+			}
+
+			ec2Ids := []*string{}
+			for _, ci := range cInstances {
+				c.cache.setContainerInstance(ci)
+				ec2Ids = append(ec2Ids, ci.Ec2InstanceId)
+			}
+
+			// Check if someone errored before continuing
+			if globErr != nil {
+				log.Debugf(globErrMsg, globErr)
+				return
+			}
+
+			// Get cluster ec2 instances
+			instances, err := c.getInstances(ec2Ids)
+			if err != nil {
+				globErr = err
+				return
+			}
+
+			for _, i := range instances {
+				c.cache.setInstance(i)
+			}
+
+			// Check if someone errored before continuing
+			if globErr != nil {
+				log.Debugf(globErrMsg, globErr)
+				return
+			}
+
+			// Get cluster tasks
+			ts, err := c.getTasks(&cluster)
+			if err != nil {
+				globErr = err
+				return
+			}
+
+			tDefIDs := []*string{}
+			for _, t := range ts {
+				c.cache.setTask(t)
+				tDefIDs = append(tDefIDs, t.TaskDefinitionArn)
+			}
+
+			// Check if someone errored before continuing
+			if globErr != nil {
+				log.Debugf(globErrMsg, globErr)
+				return
+			}
+
+			// Get cluster services
+			srvs, err := c.getServices(&cluster)
+			if err != nil {
+				globErr = err
+				return
+			}
+
+			for _, s := range srvs {
+				c.cache.setService(s)
+			}
+
+			// Check if someone errored before continuing
+			if globErr != nil {
+				log.Debugf(globErrMsg, globErr)
+				return
+			}
+
+			// Get task definitions
+			tds, err := c.getTaskDefinitions(tDefIDs, true)
+			if err != nil {
+				globErr = err
+				return
+			}
+
+			for _, t := range tds {
+				c.cache.setTaskDefinition(t)
+			}
+
+		}(*cluster)
+	}
+
+	wg.Wait()
+
+	// Check if we errored before
+	if globErr != nil {
+		log.Debugf(globErrMsg, globErr)
+		return nil, globErr
+	}
+
+	// Create the targets
+	sInstances := []*types.ServiceInstance{}
+	for _, v := range c.cache.tasks {
+		clID := aws.StringValue(v.ClusterArn)
+		cl, ok := c.cache.getCluster(clID)
+		if !ok {
+			return nil, fmt.Errorf("cluster not available: %s", clID)
+		}
+
+		ciID := aws.StringValue(v.ContainerInstanceArn)
+		ci, ok := c.cache.getContainerInstance(ciID)
+		if !ok {
+			return nil, fmt.Errorf("container not available: %s", ciID)
+		}
+
+		iID := aws.StringValue(ci.Ec2InstanceId)
+		ec2I, ok := c.cache.getIntance(iID)
+		if !ok {
+			return nil, fmt.Errorf("ec2 instance not available: %s", iID)
+		}
+
+		tdID := aws.StringValue(v.TaskDefinitionArn)
+		srv, ok := c.cache.getService(tdID)
+		if !ok {
+			return nil, fmt.Errorf("zervice not available: %s", tdID)
+		}
+
+		td, ok := c.cache.getTaskDefinition(tdID)
+		if !ok {
+			return nil, fmt.Errorf("task definition not available: %s", tdID)
+		}
+
+		t := &ecsTargetTask{
+			cluster:  cl,
+			task:     v,
+			instance: ec2I,
+			service:  srv,
+			taskDef:  td,
+		}
+
+		sis := t.createServiceInstances()
+		for _, i := range sis {
+			sInstances = append(sInstances, i)
+		}
+	}
+
+	return sInstances, globErr
+}
```

Comment:
Yeah its a memory leak (its slow, but a memory leak anyway) :/ creating a new cache on every iteration wouldn't reduce the task definition description retrieval calls. I should implement a way of invalidating that cache for example creating a new cache instance with the values of the old cache that have been hit